File: problems/basic/53/solution3.cpp
// Accepted: 87ms (46.54%), 70.18MB (15.69%)

class Solution {
public:
    int go(const vector<int>& nums, int left, int right) {
        int diff = right - left;
        if (diff == 1) {
            return nums[left];
        }
        if (diff <= 0) {
            throw std::runtime_error("not positive diff");
        }

        int mid = left + diff/2;

        int leftSum = nums[mid-1];
        int leftMax = leftSum;
        for (int i = mid-2; i >= left; --i) {
            leftSum += nums[i];
            leftMax = max(leftMax, leftSum);
        }

        int rightSum = nums[mid];
        int rightMax = rightSum;
        for (int i = mid+1; i < right; ++i) {
            rightSum += nums[i];
            rightMax = max(rightMax, rightSum);
        }

        int crossedMax = leftMax + rightMax;

        int maxSum = max(go(nums, left, mid), go(nums, mid, right));
        maxSum = max(maxSum, crossedMax);
        return maxSum;
    }

    int maxSubArray(vector<int>& nums) {
        return go(nums, 0, nums.size());
    }
};

File: problems/basic/53/solution3.py
# Accepted: 1535ms (5.01%), 45.70MB (7.93%)

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        def go(left: int = 0, right: int = len(nums)) -> int: # right is exclusive end
            diff = right - left
            if diff == 1:
                return nums[left]
            if diff <= 0:
                raise Exception

            # get left and right max to get crossed max
            mid = left + diff // 2

            leftSum = nums[mid-1]
            leftMax = nums[mid-1]
            for i in reversed(range(left, mid-1)):
                leftSum += nums[i]
                leftMax = max(leftMax, leftSum)

            rightSum = nums[mid]
            rightMax = nums[mid]
            for i in range(mid+1, right):
                rightSum += nums[i]
                rightMax = max(rightMax, rightSum)

            crossedMax = leftMax + rightMax

            return max(go(left, mid), go(mid, right), crossedMax)

        return go(0, len(nums))
